src/i_winmusic.c

```diff
@@ -598,38 +601,47 @@ static void *I_WIN_RegisterSong(void *data, int len)
         return NULL;
     }
 
-    MIDItoStream(file);
+    num_tracks = MIDI_NumTracks(song.file);
 
-    MIDI_FreeFile(file);
+    song.tracks = malloc(num_tracks * sizeof(win_midi_track_t));
+
+    for (i = 0; i < num_tracks; ++i)
+    {
+        song.tracks[i].iter = MIDI_IterateTrack(song.file, i);
+        song.tracks[i].absolute_time = 0;
+        song.tracks[i].empty = false;
+    }
 
     ResetEvent(hBufferReturnEvent);
     ResetEvent(hExitEvent);
 
-    FillBuffer();
-    StreamOut();
-
     return (void *)1;
 }
 
 static void I_WIN_UnRegisterSong(void *handle)
 {
-    FreeSong();
+    if (song.tracks)
+    {
+        int i;
+        for (i = 0; i < MIDI_NumTracks(song.file); ++i)
+        {
+            MIDI_FreeIterator(song.tracks[i].iter);
+        }
+        free(song.tracks);
+        song.tracks = NULL;
+    }
+    MIDI_FreeFile(song.file);
+    song.current_time = 0;
+    song.looping = false;
 }
 
 static void I_WIN_ShutdownMusic(void)
 {
-    MIDIHDR *hdr = &buffer.MidiStreamHdr;
     MMRESULT mmr;
 
     I_WIN_StopSong(NULL);
     I_WIN_UnRegisterSong(NULL);
 
```

Comment:
It seems complicated. Can we simply send `midiOutLongMsg()`?

Reply:
How do you know Windows actually sent it to the device other than using a callback? The error from `midiOutUnprepareHeader()` maybe? If you shutdown while the message is still in the buffer you're get an error. If you `midiStreamStop()` immediately after `midiOutLongMsg()` to prevent the error, the message won't be sent.

Reply:
Does this solution work? https://github.com/fabiangreffrath/woof/pull/816/commits/6bed6016caeeb38dd352fc8319c113796a5c1164

Reply:
No it doesn't. This works though: https://github.com/rfomin/woof/compare/winmidi_sysex...ceski-1:woof:winmidi_sysex

Reply:
> No it doesn't.

Ops, missed your message. How about this: https://github.com/fabiangreffrath/woof/pull/816/commits/fab86cf8848140cb69b2ad184a4cbcef2016122c

Reply:
Okay that works great. `buffer.position = 0` makes it really simple.